twisted/twisted/plugins/tunnel_helper_plugin.py

```diff
@@ -337,69 +365,131 @@ def cb(ds):
-        parser.add_argument('-c', '--crawl', help='Enable crawler and use the keypair specified in the given filename')
-        parser.add_argument('-j', '--json', help='Enable JSON api, which will run on the provided port number ' +
-                                                 '(only available if the crawler is enabled)', type=int)
-        parser.add_argument('-y', '--yappi', help="Profiling mode, either 'wall' or 'cpu'")
-        parser.add_help = True
-        args = parser.parse_args(sys.argv[1:])
-
-    except argparse.ArgumentError:
-        parser.print_help()
-        sys.exit(2)
-
-    socks5_port = int(args.socks5) if args.socks5 else None
-    introduce_port = int(args.introduce) if args.introduce else None
-    dispersy_port = int(args.dispersy) if args.dispersy else -1
-    crawl_keypair_filename = args.crawl
-    profile = args.yappi if args.yappi in ['wall', 'cpu'] else None
-
-    if profile:
-        yappi.set_clock_type(profile)
-        yappi.start(builtins=True)
-        logger.error("Profiling using %s time" % yappi.get_clock_type()['type'])
-
-    if crawl_keypair_filename and not os.path.exists(crawl_keypair_filename):
-        logger.error("Could not find keypair filename", crawl_keypair_filename)
-        sys.exit(1)
-
-    settings = TunnelSettings()
-
-    # For disbling anonymous downloading, limiting download to hidden services only
-    settings.min_circuits = 0
-    settings.max_circuits = 0
-
-    if socks5_port is not None:
-        settings.socks_listen_ports = range(socks5_port, socks5_port + 5)
-    else:
-        settings.socks_listen_ports = [random.randint(1000, 65535) for _ in range(5)]
-
-    settings.become_exitnode = True if args.exit in ['true'] else False
-    if settings.become_exitnode:
-        logger.info("Exit-node enabled")
-    else:
-        logger.info("Exit-node disabled")
-
-    tunnel = Tunnel(settings, crawl_keypair_filename, dispersy_port)
-    StandardIO(LineHandler(tunnel, profile))
-    tunnel.start(introduce_port)
-
-    if crawl_keypair_filename and args.json > 0:
-        cherrypy.config.update({'server.socket_host': '0.0.0.0', 'server.socket_port': args.json})
-        cherrypy.quickstart(tunnel)
-    else:
-        while tunnel.should_run:
-            time.sleep(1)
-
-    tunnel.stop()
-
-if __name__ == "__main__":
-    main(sys.argv[1:])
+class TunnelHelperServiceMaker(object):
+    implements(IServiceMaker, IPlugin)
+    tapname = "tunnel_helper"
+    description = "Tunnel Helper twistd plugin, starts a (hidden) tunnel as a service."
+    options = Options
+
+    def __init__(self):
+        """
+        Initialize the variables of this service and the logger.
+        """
+        self._stopping = False
+
+    def shutdown_process(self, shutdown_message, code=1):
+        msg(shutdown_message)
+        reactor.addSystemEventTrigger('after', 'shutdown', os._exit, code)
+        reactor.stop()
+
+    def check_options(self, options):
```

Comment:
Integrate the option validation into the Option class: https://twistedmatrix.com/documents/12.0.0/core/howto/options.html